### ISSUE TITLE
ICE compiling rustc-serialize under sccache 'failed to acquire jobserver token'

### ISSUE THREAD

**Issue by rillian** [rust-lang/rust#42867]
Failure [building Firefox](https://treeherder.mozilla.org/#/jobs?repo=try&revision=ba8913f1b02d9c01a920147fda6ffea059f3abb0&selectedJob=109654543) with rustc 1.20.0-nightly (ab5bec255 2017-06-22).

I can't reproduced on a local build with `--enable-stylo` so it's something to do with the [sccache](https://github.com/mozilla/sccache) wrapper or the build environment.

```
checking for rustc... /home/worker/workspace/build/src/rustc/bin/rustc
checking for cargo... /home/worker/workspace/build/src/rustc/bin/cargo
checking rustc version... 1.20.0-nightly
checking cargo version... 0.21.0
[...]
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: rustc 1.20.0-nightly (ab5bec255 2017-06-22) running on x86_64-unknown-linux-gnu
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'failed to acquire jobserver token: Error { repr: Os { code: 9, message: "Bad file descriptor" } }', /checkout/src/libcore/result.rs:860
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:355
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:365
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:549
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:511
   6: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:495
   7: rust_begin_unwind
             at /checkout/src/libstd/panicking.rs:471
   8: core::panicking::panic_fmt
             at /checkout/src/libcore/panicking.rs:69
   9: core::result::unwrap_failed
  10: rustc_trans::back::write::execute_work
  11: rustc_trans::back::write::run_passes
  12: rustc_driver::driver::phase_5_run_llvm_passes
  13: rustc_driver::driver::compile_input
  14: rustc_driver::run_compiler
error: Could not compile `rustc-serialize`.
Caused by:
  process didn't exit successfully: `/home/worker/workspace/build/src/sccache2/sccache /home/worker/workspace/build/src/rustc/bin/rustc --crate-name rustc_serialize /home/worker/workspace/build/src/third_party/rust/rustc-serialize/src/lib.rs --crate-type lib --emit=dep-info,link -C opt-level=2 -C debuginfo=2 -C metadata=eaa267b04fe4074c -C extra-filename=-eaa267b04fe4074c --out-dir /home/worker/workspace/build/src/obj-firefox/toolkit/library/release/deps -C linker=/home/worker/workspace/build/src/build/cargo-linker -L dependency=/home/worker/workspace/build/src/obj-firefox/toolkit/library/release/deps --cap-lints allow` (exit code: 101)
```

**Comment by rillian** [comment]
On (linux-hosted) macOS build, the same panic happens building the simd crate. It must be something to with the jobserver rather than the specific input code.

**Comment by glandium** [comment]
related to https://github.com/rust-lang/cargo/issues/1744 ?

**Comment by glandium** [comment]
The "failed to acquire jobserver token" message only appears in cargo, and comes from that issue I just mentioned.

**Comment by rillian** [comment]
The jobserver moved into the compiler's parallel support recently. See #42682.

**Comment by glandium** [comment]
Err, I had failed to update my rust tree properly, which is why I still was seeing the old code.

**Comment by alexcrichton** [comment]
Hm sounds bad! @rillian do you know if there's a way to reproduce this short of "check out gecko and build everything"? I'm trying some local smaller test cases but can't seem to get the same assertion to show up.

**Comment by rillian** [comment]
It's even harder than that. I've only seen it in the automation builds, so we need something from that environment. I couldn't reproduce with a local gecko build. Probably the best a approach is to use a taskcluster image.

I can investigate further, but it will take a few days.

**Comment by alexcrichton** [comment]
Hm ok, I'll try expanding this a bit. Yes jobserver support was recently added to both rustc and Cargo with the intention of being able to create their own and integrate with foreign jobservers. Both rustc and Cargo have the same support code, [a `jobserver` crate](https://github.com/alexcrichton/jobserver-rs).

Jobserver inheritance is specified through the `MAKEFLAGS` env var which the crate [will parse](https://github.com/alexcrichton/jobserver-rs/blob/f89e1f4c6f59f607e621f1139fd5321ca97cab4b/src/lib.rs#L196-L206) and attempt to interpret. On unix this means it'll attempt to parse the string [as file descriptors](https://github.com/alexcrichton/jobserver-rs/blob/f89e1f4c6f59f607e621f1139fd5321ca97cab4b/src/lib.rs#L441-L472) and interpret those. Note, though, that it attempts to validate that the file descriptor is indeed valid by [checking `fstat` that it's a pipe](https://github.com/alexcrichton/jobserver-rs/blob/f89e1f4c6f59f607e621f1139fd5321ca97cab4b/src/lib.rs#L605-L621). Both cargo and rustc ignore misconfigured `MAKEFLAGS` and fall back to creating their own jobserver if it looks like a funny env var.

So what's most likely to be happening here is that `MAKEFLAGS` is passed into rustc, it passes validation, and then later on it turns out it was actually invalid (for whatever reason) which causes issues. The error here, "bad file descriptor," I... don't know how that could happen! That seems to me like a file descriptor was *closed* but rustc afaik doesn't do that.

Some questions that may help narrow this down:

* Does this only happen with an `sccache` wrapper?
* If so, what revision of sccache?
* If so, is the sccache server started elsewhere?
* If not, what's the process tree look like when this step in the build process is called? Is `make` somewhere in the history? Is `cargo` the one expected to be creating jobservers? (etc)

**Comment by glandium** [comment]
sccache has a client-server model, and it's the server that does actual compilations, invoking rustc. The problem might be that the server and the client are not running in the same make environment or something related to that.

In any case, cargo/rustc should probably avoid failing in that case...

Cc: @luser

**Comment by rillian** [comment]
I [wasn't able](https://treeherder.mozilla.org/#/jobs?repo=try&revision=6a99809c11ede790ed8719c4435b7d323e28954d) to reproduce with sccache disabled.

**Comment by alexcrichton** [comment]
@rillian do you know how sccache is started in the build system? Is the server started manually way near the beginning? 

Additionally is this perhaps invoked by `make` or otherwise? Or is this run in a "standalone" fashion?

**Comment by glandium** [comment]
I /think/ the sccache server starts during the `make` run, but it is possible other compilations happen during a /separate/ `make` run. (specifically, things related to gtests would)

**Comment by glandium** [comment]
@luser would likely know better.

**Comment by rillian** [comment]
Afaict we rely on the first sccache invocation to start it. There is code to explicitly [shut down any running server](https://github.com/mozilla/gecko-dev/blob/master/build/sccache.mk#L12) before we start and after we finish. I don't see any spurious invocations of `sccache --stop-server` in the log though. Just [once](https://treeherder.mozilla.org/logviewer.html#?job_id=110039684&repo=try&lineNumber=1878) at the start of the failed run, so don't think that's it.

Is it possible traversing a `make` invocation boundary is closing the file descriptor rust (or sccache?) received from the environment?

Our builds are say they're using sccache 9155425cfc038d6a60deb50816055f4e93b93ad1, but I don't see that commit in the history. Maybe it's a fork.

**Comment by rillian** [comment]
Yes, our scache build is from https://github.com/luser/sccache/commit/9155425cfc038d6a60deb50816055f4e93b93ad1.

**Comment by alexcrichton** [comment]
Hm ok I'm still not sure how this can happen... It sounds like `sccache` is started early enough that it doesn't inherit any file descriptors. This means that `rustc` runs with `--jobserver-fds=...` but those are invalid fds. This should be [detected by the jobserver crate](https://github.com/alexcrichton/jobserver-rs/blob/f89e1f4c6f59f607e621f1139fd5321ca97cab4b/src/lib.rs#L605-L621) unless using `fstat` is just broken?

**Comment by rillian** [comment]
I never got time to investigate, but I can [still reproduce](https://treeherder.mozilla.org/logviewer.html#?job_id=113116168&repo=try&lineNumber=7478) with today's 1.20.0-nightly.

**Comment by glandium** [comment]
So, there are several things going on, but I'll just start with the one that's really relevant to cargo and rust:
- As mentioned in https://github.com/rust-lang/cargo/issues/1744#issuecomment-244246367 make is kind of a jerk, and sets the environment variables with `--jobserver-fds` even when it doesn't give out the pipes. Which means, as mentioned in that comment, the *very first* thing that any jobserver client needs to do is to check whether the file descriptors are there or not, because anything after the process startup may open file descriptors and get those numbers. This also means ICE'ing is not the best thing rustc and cargo can do in that (unfortunately) common case.
- The Firefox build system is complex. When the first call to sccache happens, which spawns the sccache server, we're in configure, running from make, running from mach, and that make instance is *without* a jobserver (essentially because the makefile is bonkers and doesn't work with parallelism). So we don't get `MAKEFLAGS` set with `--jobserver-fds`, and don't hit the rust ICE as long as we keep using that sccache server, We might be able to get away with enabling the jobserver and adding a `.NOTPARALLEL:` to that makefile, but make would still not pass the fds to configure (we'd have to prefix the configure call with a +).
- For some lucky reason (I guess that has to do with `make export` taking too long), by the time we reach the first compilation, we end up launching a new sccache server, which then inherits the `MAKEFLAGS` with `--jobserver-fds` because we're now in a make instance with a jobserver (launched from the one without from above, don't ask, I told you it was complex). But because it's not prefixed with a +, make doesn't pass the fds, and rust barfs. So in the Firefox build system, we'd need to prefix all invocations of sccache from a Makefile with a +, as well as the one for cargo. The latter would be necessary whether or not sccache is enabled for the build.

For all these reasons above, when I tried this morning with 1.20.0-nightly on a recent tree, it didn't fail, because luck. And as mentioned to @alexcrichton on irc, cargo also didn't actually use the job server, for the same reasons.

**Comment by glandium** [comment]
FWIW, other than those issues, the jobserver client support in cargo and rustc seems to work.

**Comment by alexcrichton** [comment]
Hm so I'm still confused how rustc/cargo are using a broken jobserver. There should be enough guards in place all along the way to prevent usage of a misconfigured or broken jobserver, it's expected that we get `MAKEFLAGS` without actually getting file descriptors in a sense.

The error above is EBADF returned from a call to `read`, which is very confusing to me. This means that the file descriptor is *closed* in the Rust process, which given the validation I mentioned previously that the `jobserver` crate does seems quite odd. Even if random file descriptors are leaked into a `rustc` process to fool it about the jobserver, how are the file descriptors *closed* in the `rustc` process?

**Comment by glandium** [comment]
EBADF doesn't necessarily mean the fd is closed. It means it's not a valid fd for reading. Which means it could be a valid write end of a pipe. As a matter of fact, I've seen that happen in sccache itself, but sccache doesn't try to use the fds from MAKEFLAGS, so that wasn't a problem: the process had MAKEFLAGS=--jobserver-fds=5,6, and per /proc/pid/fd, 6 and 7 were end points of the same pipe (which I presume, come from a rust channel).

Anyways, here's the likely scenario:
- Make invokes cargo with MAKEFLAGS=--jobserver-fds=n,m but doesn't pass the fds.
- Cargo starts with fd 0, 1, 2 respectively for stdin, stdout and stderr, and nothing else.
- Cargo does stuff, which opens fds, and ends up opening fds that happen to have number n and/or m
- Cargo tries to use that after checking MAKEFLAGS and kaboom.

That's a simplified scenario. The real-world problem that this issue was opened for goes one level deeper, with rustc.

The jobserver code *does* check that the fds are end points of pipes, but the fact is, by the time the fds are checked, they might well be.

Disclaimer: I haven't actually seen what fds were open in a rustc process, but this seems like a reasonable explanation. The safest way IMHO, that the jobserver code can be initialized is from a lazy_static (which doesn't even guarantee it would be first, but at least, the chances of something else opening fds are slim)

**Comment by alexcrichton** [comment]
Ah I didn't actually know you could get EBADF for the wrong end of a pipe! I guess that makes sense though, and sounds quite plausible. I'm still having difficulty constructing how this is possible to happen though. In Rust all files are atomically opened with CLOEXEC so nothing from sccache should leak into spawned processes of sccache. Additionally practically the first thing Cargo does is call [`Config::default`](https://github.com/rust-lang/cargo/blob/5a5fe12cc369561ad5ed9bf3024d892bd6356c9a/src/bin/cargo.rs#L72-L75) which'll hit [jobserver initialization](https://github.com/rust-lang/cargo/blob/5a5fe12cc369561ad5ed9bf3024d892bd6356c9a/src/cargo/util/config.rs#L47-L56). In rustc we [build s session quite early](https://github.com/rust-lang/rust/blob/b2c0707872082c890f332178f59fd02eea5b98f3/src/librustc_driver/lib.rs#L175) which'll [hit initialization](https://github.com/rust-lang/rust/blob/b2c0707872082c890f332178f59fd02eea5b98f3/src/librustc/session/mod.rs#L741-L757). 

I wonder if fds are being leaked into `sccache` itself from `make` and/or `configure`? Is there a way to check what the fds are of `sccache` itself when this happens?

It does sound like regardless that the something crate needs to be *even more* resilient to to a "misconfigured" environment. I'm not really sure what this would look like in terms of a reasonable implementation though... I wonder if `sccache` could just scrub env vars for a quick fix?

**Comment by glandium** [comment]
> I wonder if fds are being leaked into sccache itself from make and/or configure?

There are, but I think they are red herrings... as a matter of fact, when the fds from make *are* passed down (and still other fds leaked), as noted in https://github.com/rust-lang/rust/issues/42867#issuecomment-315046650 , this actually works properly.

**Comment by glandium** [comment]
I actually double checked last comment by taking the same Firefox tree from @rillian's try push, and applying bug 1367940 on top of it. https://treeherder.mozilla.org/#/jobs?repo=try&revision=b7b58c097d8c0b4340ee88c982c663655b51afcb

**Comment by staktrace** [comment]
I'm running into this same failure (ICE with 'failed to acquire jobserver token') when doing a standalone webrender build (no gecko/mozilla-central involved) with sccache. I don't get the error without sccache.

The machine it's running on is also pretty minimal - it just has the bare minimum of things needed to build webrender. Latest stable rust/cargo (1.20.0) and latest sccache (0.2.1). The build is run on the machine through taskcluster integration, but the machine is in a server farm we rented from macstadium and that we're managing manually. You can see the full taskcluster log at https://tools.taskcluster.net/groups/BqgBtyCvQUqgDaQAJglrUA/tasks/Q8UMqP9TR7-va2ztvs-lSw/runs/0/logs/public%2Flogs%2Flive.log

It would be great to get this resolved as we want to use this setup for webrender CI and being able to use sccache means we'll get much faster turnaround.

**Comment by alexcrichton** [comment]
@staktrace hm that's a very curious error... The OS error there is "Resource temporarily unavailable" with code 35 which seems to correspond to EDEADLK. Apparently calling `read` led to `EDEADLK`?

Are you running `sccache` manually ahead of time? Or is it just getting spawned naturally through the first invocation?

**Comment by staktrace** [comment]
@alexcrichton  It is getting spawned naturally via the first invocation.

**Comment by alexcrichton** [comment]
Oh I notice now the error is on OSX, not on Linux, so 35 is EAGAIN which makes more sense! (EDEADLK seemed crazy...)

So I believe that EAGAIN primarily has to do with nonblocking mode, which I think means that something is setting these file descriptors to nonblocking mode, which would certainly cause problems! As to *what* is setting them to nonblocking mode... is a good question.

@staktrace can you reproduce this locally by hand? Or is this only reproducible going through taskcluster?

**Comment by staktrace** [comment]
So I can reproduce this by hand on the same machine, if I run the following:
```bash
  export RUST_BACKTRACE=1
  export RUSTFLAGS='--deny warnings'
  export PKG_CONFIG_PATH="/usr/local/opt/zlib/lib/pkgconfig:$PKG_CONFIG_PATH"
  export RUSTC_WRAPPER=sccache
  cd webrender_api/
  cargo test --verbose --features "ipc"
```

I also tried just running `sccache --start-server` and that resulted in this:

```
Starting sccache server...
error: Server startup error
caused by: Server startup failed: 1
```

So maybe that's the problem? Do I need to start the server with root privileges? I did try to lock down permissions on the user account that this is running under, for security purposes, but I'm not sure if anything I did would have caused this. Most of the locking down I did was through the macOS GUI for user accounts which doesn't seem like it would affect anything sccache would want to do.

**Comment by alexcrichton** [comment]
Weird and.. fascinating! When you run via cargo sccache is starting... something because rustc is being invoked and then failing. When you run by hand though sccache can't even start. I... have no idea what's going on.

Can you try getting some logs out of sccache to see where the error is going? I think you'll have to do something like --

```
SCCACHE_ERROR_LOG=tmp.log RUST_LOG=debug sccache --start-server
```

and then I think `tmp.log` may have some contents?

**Comment by staktrace** [comment]
I ran the above command but nothing got written to tmp.log, but there was some output to the console:

```
$ SCCACHE_ERROR_LOG=tmp.log RUST_LOG=debug sccache --start-server
Starting sccache server...
DEBUG:tokio_core::reactor: adding a new I/O source
DEBUG:tokio_core::reactor: added a timeout: 0
DEBUG:tokio_core::reactor: consuming notification queue
DEBUG:tokio_core::reactor: scheduling direction for: 0
DEBUG:tokio_core::reactor: blocking
DEBUG:tokio_core::reactor: consuming notification queue
DEBUG:tokio_core::reactor: updating a timeout: 0
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 0, nanos: 4040 }
DEBUG:tokio_core::reactor: loop time - Instant { t: 938828749006035 }
DEBUG:tokio_core::reactor: consuming notification queue
DEBUG:tokio_core::reactor: loop process - 1 events, Duration { secs: 0, nanos: 15762 }
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 0, nanos: 630 }
DEBUG:tokio_core::reactor: loop time - Instant { t: 938828749025008 }
DEBUG:tokio_core::reactor: loop process - 0 events, Duration { secs: 0, nanos: 4236 }
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 5, nanos: 4990786 }
DEBUG:tokio_core::reactor: loop time - Instant { t: 938833754022992 }
DEBUG:tokio_core::reactor: notifying a task handle
DEBUG:tokio_core::reactor: loop process - 0 events, Duration { secs: 0, nanos: 121377 }
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 0, nanos: 33354 }
DEBUG:tokio_core::reactor: loop time - Instant { t: 938833754185221 }
DEBUG:tokio_core::reactor: loop process - 1 events, Duration { secs: 0, nanos: 9075 }
DEBUG:tokio_core::reactor: consuming notification queue
DEBUG:tokio_core::reactor: scheduling direction for: 0
DEBUG:tokio_core::reactor: blocking
DEBUG:tokio_core::reactor::timeout_token: cancel timeout 0
DEBUG:tokio_core::reactor: consuming notification queue
DEBUG:tokio_core::reactor: cancel a timeout: 0
error: Timed out waiting for server startup
```

I noticed the output was a little different from last time so I tried starting it again just for kicks, and this time it started!

```
$ sccache --start-server
Starting sccache server...
$ ps aux | grep sccache
tcworker        33579   0.0  0.0  2499724   2076   ??  S     5:23am   0:00.00 /Users/tcworker/.cargo/bin/sccache
tcworker        33582   0.0  0.0  2432804   1912 s000  S+    5:23am   0:00.00 grep sccache
```

Really weird...

**Comment by staktrace** [comment]
Ah. I think the error I got before with `sccache --start-server` was a red herring. I think the sccache server was already running at that point, started from the previous build I had attempted. I can reproduce that error output if I try starting the server when it's already running in the background.

At this point I can start and stop the server without errors, but still have the original ICE when running the build commands for webrender with RUSTC_WRAPPER=sccache.

**Comment by alexcrichton** [comment]
I... still have no idea how to explain this. 

Using sccache sort of fundamentally can't work with how jobservers are designed, but rustc *should* have a check to prevent it from using bad fds. This means that somehow bad fds are leaking in, and I have no idea how to explain that.

In any case none of this is really supposed to work anyway (sccache working with jobservers) so I think we need to either:

1. Remove any jobserver-related env vars from spawned processes in sccache
2. Configure sccache with its own jobserver which is passed to spawned processes.

I'm personally a little in favor of (2)

**Comment by aneeshusa** [comment]
I haven't looked into this at all yet, but wanted to mention I'm getting similar errors when trying to [enable this on Servo CI](https://github.com/servo/servo/pull/17171).

**Comment by alexcrichton** [comment]
I've opened https://github.com/mozilla/sccache/pull/185 with my personal preferred strategy of adding a jobserver to sccache.

**Comment by whitslack** [comment]
> So I believe that EAGAIN primarily has to do with nonblocking mode, which I think means that something is setting these file descriptors to nonblocking mode, which would certainly cause problems! As to _what_ is setting them to nonblocking mode... is a good question.

Make [sets](http://git.savannah.gnu.org/cgit/make.git/commit/?id=b552b05251980f693c729e251f93f5225b400714) its jobserver pipe file descriptors to non-blocking mode when it's configured to use `pselect`.

**Comment by luser** [comment]
Yeah, the `EAGAIN` in cargo is almost 100% reproducible using gmake master to build Firefox. Changing that `#ifdef HAVE_PSELECT` to `#if 0` makes it go away.

**Comment by alexcrichton** [comment]
@whitslack excellent point! I've got a fix for your and @luser's issue at https://github.com/alexcrichton/jobserver-rs/pull/2

**Comment by jrmuizel** [comment]
Any update on this?

**Comment by alexcrichton** [comment]
@luser want to take a look at https://github.com/mozilla/sccache/pull/185 and see if you agree with it?

**Comment by luser** [comment]
I finally merged @alexcrichton's change into sccache. If someone wants to test to see if this indeed fixes the issues in Servo CI and elsewhere I'd be happy to tag a new sccache release to get new binaries.

**Comment by aneeshusa** [comment]
@luser I'd definitely be interested in trying out the latest in Servo CI! If you cut a new release (or even a pre-release/beta since we're testing) and we can grab some new binaries, we can give it a whirl.

**Comment by staktrace** [comment]
FWIW my STR from https://github.com/rust-lang/rust/issues/42867#issuecomment-331606567 seem to be fixed with the latest published sccache (0.2.6)

**Comment by alexcrichton** [comment]
Hurray!
